lib/spack/spack/dependency.py

```diff
@@ -0,0 +1,133 @@
+##############################################################################
+# Copyright (c) 2013-2017, Lawrence Livermore National Security, LLC.
+# Produced at the Lawrence Livermore National Laboratory.
+#
+# This file is part of Spack.
+# Created by Todd Gamblin, tgamblin@llnl.gov, All rights reserved.
+# LLNL-CODE-647188
+#
+# For details, see https://github.com/llnl/spack
+# Please also see the NOTICE and LICENSE files for our notice and the LGPL.
+#
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU Lesser General Public License (as
+# published by the Free Software Foundation) version 2.1, February 1999.
+#
+# This program is distributed in the hope that it will be useful, but
+# WITHOUT ANY WARRANTY; without even the IMPLIED WARRANTY OF
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the terms and
+# conditions of the GNU Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public
+# License along with this program; if not, write to the Free Software
+# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
+##############################################################################
+"""Data structures that represent Spack's dependency relationships.
+"""
+from six import string_types
+
+import spack
+
+
+#: The types of dependency relationships that Spack understands.
+all_deptypes = ('build', 'link', 'run')
+
+#: Default dependency type if none is specified
+default_deptype = ('build', 'link')
+
+
+def canonical_deptype(deptype):
+    """Convert deptype to a canonical sorted tuple, or raise ValueError.
+
+    Args:
+        deptype (str or list or tuple): string representing dependency
+            type, or a list/tuple of such strings.  Can also be the
+            builtin function ``all`` or the string 'all', which result in
+            a tuple of all dependency types known to Spack.
+    """
+    if deptype in (None, 'all', all):
+        return all_deptypes
+
+    elif isinstance(deptype, string_types):
+        if deptype not in all_deptypes:
+            raise ValueError('Invalid dependency type: %s' % deptype)
```

Comment:
I think when a `ValueError` occurs, it's generally because there is an error in the program, not something that would need to be handled programmatically or that you'd need an exception handler for.  The only place I've seen `ValueError` even caught is if someone's trying to construct an int from a string or something similar, and generally in that case you juts make a really narrow try/catch statement.

For things that the caller would actually want to handle (like network errors, or issues with packages) I think wrapping in a SpackError makes a lot more sense.